sfge/src/main/java/com/salesforce/graph/ops/expander/PathExpansionRegistry.java

```diff
@@ -0,0 +1,119 @@
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+
+/**
+ * Optimization to maintain lower heap space usage while expanding paths. See {@link Registry} for
+ * information on how the optimization works.
+ *
+ * <p>This registry implementation holds data for {@link ApexPathCollapser}, {@link
+ * ApexPathExpander}, and {@link ForkEvent} instances. Collections that need to hold instances of
+ * these types, can hold their {@link Long} Ids instead, and use the corresponding <code>lookup()
+ * </code> method to fetch the instance from registry.
+ *
+ * <p>{@link PathExpansionRegistry} uses instances of {@link AbstractRegistryData} rather than
+ * {@link ThreadLocal} so that these hashmaps only grow as needed per path expansion of an ApexPath.
+ * Once cleared, all heap usage will be released back. This makes sure:
+ * <ul>
+ *     <li>1. Registry data lives for a shorted duration of time and doesn't move to Old Gen of heap space until later. This allows
+ *  earlier clean up from heap space when they are not used.</li>
+ *     <li>2. Internally-used HashMaps can start
+ *  with smaller sizes and can grow as needed, as opposed to starting with a much larger HashMap that
+ *  was needed in a previous path expansion.</li>
+ * </ul>
+ *
+ *
+ * <p>Create a new instance of this registry when a path expansion for an ApexPath is initiated at
+ * {@link ApexPathExpanderUtil#expand(GraphTraversalSource, ApexPath, ApexPathExpanderConfig)}.
+ * Clear the registry when the operation completes to avoid accidental memory leaks.
+ */
+public class PathExpansionRegistry extends Registry {
+
+    private static final Map<Class<? extends Indexable>, Supplier> REGISTRY_SUPPLIER =
+            ImmutableMap.of(
+                    ApexPathCollapser.class, () -> new PathCollapserRegistryData(),
+                    ForkEvent.class, () -> new ForkEventRegistryData(),
+                    ApexPathExpander.class, () -> new ApexPathExpanderRegistryData());
+
+    @Override
+    protected Map<Class<? extends Indexable>, Supplier> getRegistrySupplier() {
+        return REGISTRY_SUPPLIER;
+    }
+
+    /** Registry data structure to hold {@link ApexPathCollapser}. */
+    private static class PathCollapserRegistryData extends AbstractRegistryData<ApexPathCollapser> {
+        // Nothing new to add
+    }
+
+    /** Registry data structure to hold {@link ForkEvent}. */
+    private static class ForkEventRegistryData extends AbstractRegistryData<ForkEvent> {
+        // Nothing new to add
+    }
+
+    /** Registry data structure to hold {@link ApexPathExpander}. */
+    private static class ApexPathExpanderRegistryData
+            extends AbstractRegistryData<ApexPathExpander> {
+        // Nothing new to add
+    }
+
+    ///// ApexPathCollapser registry methods//////
+
+    public void registerApexPathCollapser(ApexPathCollapser apexPathCollapser) {
```

Comment:
It feels weird for all of these separate functions to exist for each subtype, instead of just having one method that covers any subtype. Is that possible to do?

Reply:
I'm hoping to reduce the possibility of passing in an incorrect first parameter. While making changes, I dealt with passing in the wrong class and had to debug it. These redundant methods fortify the entry to use the registry.

Reply:
Could you use generics? E.g., `<T extends Indexable>` and `.getClass()`?